// react-super-responsive-table/dist/cjs/components/allts.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeaderContext = void 0;
exports.HeaderProvider = HeaderProvider;
exports.Table = TableWithHeaderProvider;
exports.Tbody = Tbody;
exports.Td = Td;
exports.TdInner = TdInner;
exports.Th = Th;
exports.Thead = Thead;
exports.Tr = Tr;
exports.TrInner = TrInner;
var react_1 = __importStar(require("react"));
var omit = function (obj, omitProps) {
    return Object.keys(obj)
        .filter(function (key) { return !omitProps.includes(key); })
        .reduce(function (returnObj, key) {
        var _a;
        return __assign(__assign({}, returnObj), (_a = {}, _a[key] = obj[key], _a));
    }, {});
};
var allowed = function (props) {
    return omit(props, ['inHeader', 'columnKey', 'headers', 'forwardedRef']);
};
var HeaderContext = (0, react_1.createContext)(undefined);
exports.HeaderContext = HeaderContext;
function HeaderProvider(_a) {
    var children = _a.children;
    var _b = (0, react_1.useState)([]), headers = _b[0], setHeaders = _b[1];
    var contextValue = (0, react_1.useMemo)(function () { return ({ headers: headers, setHeaders: setHeaders }); }, [headers]);
    return (react_1.default.createElement(HeaderContext.Provider, { value: contextValue }, children));
}
var Table = (0, react_1.forwardRef)(function (_a, ref) {
    var className = _a.className, props = __rest(_a, ["className"]);
    var classes = "".concat(className || '', " responsiveTable");
    return (react_1.default.createElement("table", __assign({ "data-testid": "table" }, allowed(props), { className: classes, ref: ref })));
});
Table.displayName = 'Table';
function TableWithHeaderProvider(props) {
    return (react_1.default.createElement(HeaderProvider, null,
        react_1.default.createElement(Table, __assign({}, props))));
}
function Tbody(props) {
    return react_1.default.createElement("tbody", __assign({ "data-testid": "tbody" }, allowed(props)));
}
function Td(props) {
    var context = (0, react_1.useContext)(HeaderContext);
    if (!context) {
        throw new Error('Td must be used as a child of the Table component');
    }
    var headers = context.headers;
    return react_1.default.createElement(TdInner, __assign({}, props, { headers: headers }));
}
function TdInner(props) {
    var headers = props.headers, children = props.children, columnKey = props.columnKey, className = props.className, colSpan = props.colSpan;
    var classes = "".concat(className || '', " pivoted");
    if (colSpan) {
        return (react_1.default.createElement("td", __assign({ "data-testid": "td" }, allowed(props))));
    }
    return (react_1.default.createElement("td", __assign({ "data-testid": "td" }, allowed(props), { className: classes }),
        react_1.default.createElement("div", { "data-testid": "td-before", className: "tdBefore" }, headers[columnKey]), children !== null && children !== void 0 ? children : react_1.default.createElement("div", null, "\u00A0")));
}
function Th(props) {
    return (react_1.default.createElement("th", __assign({ "data-testid": "th" }, allowed(props))));
}
function Thead(props) {
    var children = props.children;
    return (react_1.default.createElement("thead", __assign({ "data-testid": "thead" }, allowed(props)), react_1.default.cloneElement(children, { inHeader: true })));
}
function Tr(props) {
    var context = (0, react_1.useContext)(HeaderContext);
    if (!context) {
        throw new Error('Tr must be used as a child of the Table component');
    }
    var headers = context.headers;
    return react_1.default.createElement(TrInner, __assign({}, props, { headers: headers }));
}
function TrInner(_a) {
    var children = _a.children, inHeader = _a.inHeader, props = __rest(_a, ["children", "inHeader"]);
    var context = (0, react_1.useContext)(HeaderContext);
    if (!context) {
        throw new Error('TrInner must be used as a child of the Table component');
    }
    var setHeaders = context.setHeaders;
    (0, react_1.useEffect)(function () {
        if (inHeader) {
            var newHeaders = react_1.default.Children.map(children, function (child) { return child === null || child === void 0 ? void 0 : child.props.children; });
            setHeaders(newHeaders);
        }
    }, [children, inHeader, setHeaders]);
    return (react_1.default.createElement("tr", __assign({ "data-testid": "tr" }, allowed(props)), children &&
        react_1.default.Children.map(children, function (child, i) {
            return child
                ? react_1.default.cloneElement(child, {
                    key: i,
                    columnKey: i,
                })
                : null;
        })));
}
exports.default = TrInner;
